File: backend/handlers/logs.go
package handlers

import (
	"fmt"
	"log"
	"net/http"
	"strconv"
	"time"

	"quicklens/backend/db"
	"quicklens/backend/models"
)

func SearchLogsHandler(w http.ResponseWriter, r *http.Request) {
	page, _ := strconv.Atoi(r.URL.Query().Get("page"))
	perPage, _ := strconv.Atoi(r.URL.Query().Get("per_page"))
	if page < 1 {
		page = 1
	}
	if perPage < 1 || perPage > 100 {
		perPage = 50
	}
	offset := (page - 1) * perPage

	// Filters
	where := "WHERE type = 'llm'"
	args := make([]interface{}, 0)

	if model := r.URL.Query().Get("model"); model != "" {
		where += " AND model_id = ?"
		args = append(args, model)
	}
	if status := r.URL.Query().Get("status"); status != "" {
		where += " AND status = ?"
		args = append(args, status)
	}
	if minLatencyStr := r.URL.Query().Get("min_latency"); minLatencyStr != "" {
		if minLatency, err := strconv.ParseInt(minLatencyStr, 10, 64); err == nil {
			where += " AND duration_ms >= ?"
			args = append(args, minLatency)
		}
	}
	if maxLatencyStr := r.URL.Query().Get("max_latency"); maxLatencyStr != "" {
		if maxLatency, err := strconv.ParseInt(maxLatencyStr, 10, 64); err == nil {
			where += " AND duration_ms <= ?"
			args = append(args, maxLatency)
		}
	}
	if from := r.URL.Query().Get("from"); from != "" {
		where += " AND started_at >= ?"
		args = append(args, from)
	}
	if to := r.URL.Query().Get("to"); to != "" {
		where += " AND started_at <= ?"
		args = append(args, to)
	}

	query := fmt.Sprintf(`
		SELECT id, trace_id, model_id, provider, status, duration_ms,
		       prompt_tokens, completion_tokens, input, output, error_message, started_at
		FROM spans
		%s
		ORDER BY started_at DESC
		LIMIT ? OFFSET ?
	`, where)

	// Append pagination args
	queryArgs := append(args, perPage, offset)

	rows, err := db.DB.Query(query, queryArgs...)
	if err != nil {
		log.Printf("Failed to query log spans: %v", err)
		WriteError(w, http.StatusInternalServerError, "Failed to query logs")
		return
	}
	defer rows.Close()

	result := make([]models.LogEntry, 0)
	for rows.Next() {
		var entry models.LogEntry
		var input, output string
		var startedAt time.Time

		err := rows.Scan(
			&entry.ID, &entry.TraceID, &entry.ModelName, &entry.Provider, &entry.Status, &entry.DurationMs,
			&entry.PromptTokens, &entry.CompletionTokens, &input, &output, &entry.ErrorMessage, &startedAt,
		)
		if err != nil {
			log.Printf("Failed to scan log span: %v", err)
			continue
		}
		entry.SpanID = entry.ID
		entry.InputPreview = truncate(input, 200)
		entry.OutputPreview = truncate(output, 200)
		entry.CreatedAt = &startedAt
		result = append(result, entry)
	}

	WriteJSON(w, http.StatusOK, result)
}

File: frontend/src/lib/api/logs.ts
import { apiGet } from './client';

export interface LogEntry {
	id: string;
	trace_id?: string;
	span_id?: string;
	model_name?: string;
	provider?: string;
	status: 'ok' | 'error';
	duration_ms: number;
	prompt_tokens: number;
	completion_tokens: number;
	input_preview: string;
	output_preview: string;
	error_message?: string;
	created_at: string;
}

export interface LogSearchParams {
	model?: string;
	status?: string;
	min_latency?: number;
	max_latency?: number;
	search?: string;
	page?: number;
	limit?: number;
}

export interface LogSearchResult {
	items: LogEntry[];
	total: number;
	page: number;
	limit: number;
}

export function searchLogs(params?: LogSearchParams): Promise<LogEntry[]> {
	const searchParams = new URLSearchParams();
	if (params) {
		if (params.page !== undefined) searchParams.set('page', String(params.page));
		if (params.limit !== undefined) searchParams.set('per_page', String(params.limit));
		if (params.model !== undefined && params.model !== '') searchParams.set('model', params.model);
		if (params.status !== undefined && params.status !== '') searchParams.set('status', params.status);
		if (params.min_latency !== undefined) searchParams.set('min_latency', String(params.min_latency));
		if (params.max_latency !== undefined) searchParams.set('max_latency', String(params.max_latency));
		if (params.search !== undefined && params.search !== '') searchParams.set('search', params.search);
	}
	const query = searchParams.toString();
	return apiGet<LogEntry[]>(`/api/v1/logs${query ? `?${query}` : ''}`);
}
